custom/onse/tasks.py

```diff
@@ -0,0 +1,298 @@
+        case = case_accessors.get_case(case_id)
+        if not case.external_id:
+            # This case is not mapped to a facility in DHIS2.
+            continue
+        case_block = CaseBlock(
+            case_id=case.case_id,
+            external_id=case.external_id,
+            case_type=CASE_TYPE,
+            case_name=case.name,
+            update={},
+        )
+        yield case_block
+
+
+def set_case_updates(
+    connection_settings: ConnectionSettings,
+    case_blocks: Iterable[CaseBlock]
+) -> Iterable[CaseBlock]:
+    """
+    Fetch all the data elements for the data set, because we don't know
+    in advance what category option combos to query for
+    """
+    # e.g. https://play.dhis2.org/dev/api/dataValueSets?orgUnit=jNb63DIHuwU
+    #   &period=2020Q2
+    #   &dataSet=QX4ZTUbOt3a  <-- The data set of the data element
+    # not https://play.dhis2.org/dev/api/dataValues?ou=jNb63DIHuwU
+    #   &pe=2020Q2
+    #   &de=gQNFkFkObU8  <-- The data element we want
+    #   &co=L4P9VSgHkF6  <-- We don't know this, and without it we get a 409
+    requests = connection_settings.get_requests()
+    endpoint = '/api/dataValueSets'
+    params = {'period': get_last_quarter()}
+    for case_block in case_blocks:
+        params['orgUnit'] = case_block.external_id
+        data_set_cache = {}
+        for case_property, (data_element, data_set) in CASE_PROPERTY_MAP.items():
+            if data_set not in data_set_cache:
+                params['dataSet'] = data_set
+                response_json = requests.get(endpoint, params).json()
+                data_set_cache[data_set] = response_json.get('dataValues', None)
+
+            if data_set_cache[data_set] is None:
+                # No data for this facility. `None` = "We don't know"
+                case_block.update[case_property] = None
+            else:
+                value = 0
+                for data_value in data_set_cache[data_set]:
+                    if data_value['dataElement'] == data_element:
+                        value += int(data_value['value'])
+                case_block.update[case_property] = value
+        yield case_block
+
+
+def get_last_quarter(today: Optional[date] = None) -> str:
+    """
+    Returns the last quarter in  DHIS2 web API `period format`_.
+    e.g. "2004Q1"
+
+    .. _period format: https://docs.dhis2.org/master/en/developer/html/webapi_date_perid_format.html
+    """
+    if today is None:
+        today = date.today()
+    year = today.year
+    last_quarter = (today.month - 1) // 3
+    if last_quarter == 0:
+        year -= 1
+        last_quarter = 4
+    return f"{year}Q{last_quarter}"
+
+
+def save_cases(case_blocks):
+    today = date.today().isoformat()
+    for chunk in chunked(case_blocks, 1000, collection=list):
+        submit_case_blocks(
+            [cb.as_text() for cb in chunk],
+            DOMAIN,
+            xmlns='http://commcarehq.org/dhis2-import',
```

Comment:
keeping things clean, I see :stuck_out_tongue_closed_eyes:

Reply:
Is this a good idea? I know this doesn't mean anything really, but we do have the openrosa convention currently, and I'm not convinced we don't assume that in places in HQ.

Reply:
Using non-openrosa XMLNS happens quite a lot already, usually to indicate the origin or purpose of the form, so I think it should be OK.

e.g.
```python
COMMTRACK_REPORT_XMLNS = 'http://commcarehq.org/ledger/v1'
DEVICE_LOG_XMLNS = 'http://code.javarosa.org/devicereport'
COMMTRACK_SUPPLY_POINT_XMLNS = 'http://commtrack.org/supply_point'
SMS_XMLNS = 'http://commtrack.org/sms_submission'
XMLNS_OPENMRS = 'http://commcarehq.org/openmrs-integration'
```